_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,359 @@
+The final step consists of writing the end transaction markers to each involved partition, and the `CompleteCommit` or `CompleteAbort` control record.
+
+![txn log](/assets/images/posts/2023-04-14-kafka-txn-log.png)
+
+The last stable offset (LSO) is the offset in a user partition such that all lower offsets have been decided and it is always present.
+The first unstable offset (FUO) is the earliest offset in a user partition that is part of the ongoing transaction, if present.
+
+The LSO is equal to the FUO if it's lower than the HW, otherwise it's the HW.
+We have that `LSO <= HW <= LEO`, which can be the same offset when there is no open transaction and all partitions are in-sync.
+
+#### Performance
+
+There are a few drawbacks when using Kafka transactions that developers should be aware of.
+
+Firstly, enabling EOS has an effect on throughput, which can be significant.
+Each producer can only process transactions sequentially, so you might need multiple transactional producers, which can have a knock-on effect on the rest of the application.
+This is where frameworks can provide useful abstractions over multiple producer instances.
+
+Then, transactions can add some latency to an application due to the write amplification required to ensure EOS.
+While this may not be a significant issue for some applications, those that require real-time processing may be affected.
+
+Each transaction in Kafka incurs some overhead that is independent of the number of messages involved.
+This overhead includes the following:
+
+- A small number of Remote Procedure Calls (RPCs) to register the partitions with the coordinator.
+  These RPCs can be batched together to reduce overhead.
+- A marker that must be written to each partition when the transaction is completed.
+- A few state change records that are appended to the internal transaction state log.
+
+Most of this latency is on the producer side, where the transactions API is implemented.
+The consumer is only be able to fetch up to the LSO, but it has to buffer records in memory until the commit marker is observed, so you have increased memory usage.
+
+The performance degradation can be significant when having short transaction commit intervals.
+Increasing the transaction duration also increases the end-to-end latency and may require some additional tuning, so it's a matter of tradeoff.
+
+#### Hanging transactions
+
+Hanging transactions have a missing or out-of-order control record and may be caused by delayed messages due to a network issue ([KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).
+This issue is rare, but it's important to be aware and set alerts, because the consequences can impact cluster availability.
+
+The transaction coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
+This mechanism does not work for hanging transactions because they are marked as completed.
```

Comment:
I wonder if a little more detail helps, wdyt?

```suggestion
This mechanism does not work for hanging transactions because from the coordinator's point of view the transaction was completed and a transaction marker written (and no longer needs to be timed out). But from the partition leader's point of view there is a data record for a transaction after the commit/abort marker, which is indistinguishable from the start of a new transaction.
```